File: luajit/bufimg.lua
local ffi = require("ffi")
local bit = require("bit")

local M = {    
}


ffi.cdef[[
void *fopen (const char * __filename,
		    const char * __modes);
int fclose (void *__stream);
size_t fread (void *__restrict __ptr, size_t __size,
		     size_t __n, void *__restrict __stream);
size_t fwrite (const void *__restrict __ptr, size_t __size,
		      size_t __n, void *__restrict __s);

]]

-- typedef int32_t[4] rgba_44;

local BufImg = {}
M.BufImg = BufImg

function BufImg.create()
    local inst = {}
    local m = {__index=BufImg}
    setmetatable(inst, m)
    return inst
end

function BufImg:isInBounds(x,y)
    return ((x > 0) and (x < self.w) and (y > 0) and (y < self.h))
end


function BufImg:ofs(x,y)
    return self.w*y*4 + x*4
end

function BufImg:load(fn)
    if fn == nil then error("can't open nil") end

    local C = ffi.C
    local f = C.fopen(fn, "rb")
    
    local wh = ffi.new("int32_t[1]")
    C.fread(wh, 4, 1, f)
    self.w = wh[0]
    C.fread(wh, 4, 1, f)
    self.h = wh[0]
    self.sz = self.w * self.h * 4
    self.buf = ffi.new("uint8_t[?]", self.sz)
    C.fread(self.buf, self.sz, 1, f)
    C.fclose(f)
    
    print("loaded image "..self.w.."x"..self.h.." sz = "..self.sz)
end

function BufImg:save(fn)
    if fn == nil then error("can't open nil") end
        
    local f = ffi.C.fopen(fn, "wb")
    local C = ffi.C
    local wh = ffi.new("int32_t[1]")
    wh[0] = self.w
    C.fwrite(wh, 4, 1, f)
    wh[0] = self.h
    C.fwrite(wh, 4, 1, f)
    C.fwrite(self.buf, self.sz, 1, f)
    C.fclose(f)
end

function BufImg:setPixelAt(byteofs, rgba)
    for i = 0, 3 do
        self.buf[byteofs + i] = rgba[i]
    end
end

function BufImg:getPixelAt(byteofs, rgba)
    for i = 0, 3 do
        rgba[i] = self.buf[byteofs + i]
    end
end

--[[

ffi.cdef
typedef struct { uint8_t red, green, blue, alpha; } trgba;





function bytes_to_int(b1, b2, b3, b4)
    if not b4 then error("need four bytes to convert to int",2) end
    local n = b1 + b2*256 + b3*65536 + b4*16777216
    n = (n > 2147483647) and (n - 4294967296) or n
    return n
end

function int_to_bytes(i)
    return {band(i, 255), band(bshr(i, 8), 255), band(bshr(i, 16), 255), band(bshr(i, 24), 255)}  
end
    
function buf_img_load(fn)
    
    local f = io.open(fn, "rb")
    
    local bytes = f:read(4)
    local w = bytes_to_int(bytes[1], bytes[2], bytes[3], bytes[4])
    local bytes = f:read(4)
    local h = bytes_to_int(bytes[1], bytes[2], bytes[3], bytes[4])
    
    local img_sz = w*h*4
    local buf = ffi.new("trgba[?]", img_sz)
    
    local bytes = f:read("*all")
    for i =1, #bytes do
        buf[i-1] = bytes[i]
    end
    
    f:close()    
    
    return {w = w, h = h, buf = buf, img_sz = img_sz}
end
    

function buf_img_save(img, fn)
    local f = io.open(fn, "wb")
    local bytes = int_to_bytes(img.w)
    f:write(bytes)
    local bytes = int_to_bytes(img.h)
    f:write(bytes)
    for i = 0, img.img_sz-1 do
       f:write(img.buf[i]) 
    end
    
    
    f:close()
end

--]]

-- export module:
return M

File: julia/BufImg.jl
module BufImg

export Img, load, save, ofs, Rgba, set_pixel_at, get_pixel_at, sz, is_in_bounds
export WidthType, ChannelType

const WidthType = Int
const ChannelType = Int32

struct Img
    w::WidthType
    h::WidthType
    buf::Array{UInt8}

    function Img(w::WidthType, h::WidthType) 
        new(w, h, Base.zeros(UInt8, sz(w,h)))
    end

    function Img(w::WidthType, h::WidthType, buf) 
        new(w, h, buf)
    end
end

struct Rgba
    r::ChannelType
    g::ChannelType
    b::ChannelType
    a::ChannelType

    function Rgba(r,g,b,a)
        new(ChannelType(r), ChannelType(g), ChannelType(b), ChannelType(a))
    end
end

function load(fn::String)::Img
    w::WidthType = 0
    h::WidthType = 0
    buf::Array{UInt8} = []
    open(fn) do io
        w = read(io, Int32)
        h = read(io, Int32)
        bufsz = sz(w,h)
        buf = Array{UInt8}(undef, bufsz)
        read!(io, buf)
    end
    Img(w,h,buf)
end

function save(img:: Img, fn::String)
    open(fn, "w") do io
        write(io, Int32(img.w))
        write(io, Int32(img.h))
        write(io, img.buf)
    end
end


function sz(w, h)::Int
    w*h*4
end

function sz(img::Img)::Int
    sz(img.w, img.h)
end

function ofs(img::Img, x::WidthType, y::WidthType)::Int
    1+(y*img.w + x)*4
end

function set_pixel_at(img::Img, byte_ofs::Int, color::Rgba)
    img.buf[byte_ofs] = convert(UInt8, color.r)
    img.buf[byte_ofs+1] = convert(UInt8, color.g)
    img.buf[byte_ofs+2] = convert(UInt8, color.b)
    img.buf[byte_ofs+3] = convert(UInt8, color.a)
end

function get_pixel_at(img::Img, byte_ofs::Int)::Rgba
    Rgba(img.buf[byte_ofs], img.buf[byte_ofs+1],img.buf[byte_ofs+2],img.buf[byte_ofs+3])
end

function is_in_bounds(img::Img, x::WidthType, y::WidthType)::Bool
    x >= 0 && y >= 0 && x < img.w && y < img.h
end


end